src/iter/unzip.rs

```diff
@@ -0,0 +1,236 @@
+use super::internal::*;
+use super::*;
+
+
+/// Unzips the items of a parallel iterator into a pair of arbitrary
+/// `ParallelExtend` containers.
+pub fn unzip<I, A, B, FromA, FromB>(pi: I) -> (FromA, FromB)
+    where I: ParallelIterator<Item = (A, B)>,
+          FromA: Default + ParallelExtend<A>,
+          FromB: Default + ParallelExtend<B>,
+          A: Send,
+          B: Send
+{
+    let mut a = FromA::default();
+    let mut b = FromB::default();
+    {
+        // We have no idea what the consumers will look like for these
+        // collections' `par_extend`, but we can intercept them in our own
+        // `drive_unindexed`.  Start with the left side, type `A`:
+        let iter = UnzipA {
+            base: pi,
+            b: &mut b,
+        };
+        a.par_extend(iter);
+    }
+    (a, b)
+}
+
+/// Unzip an `IndexedParallelIterator` into two arbitrary `Consumer`s.
+pub fn unzip_indexed<I, A, B, CA, CB>(pi: I, left: CA, right: CB) -> (CA::Result, CB::Result)
+    where I: IndexedParallelIterator<Item = (A, B)>,
+          CA: Consumer<A>,
+          CB: Consumer<B>,
+          A: Send,
+          B: Send
+{
+    let consumer = UnzipConsumer {
+        left: left,
+        right: right,
+    };
+    pi.drive(consumer)
+}
+
+/// Unzip a `ParallelIterator` into two arbitrary `UnindexedConsumer`s.
+pub fn unzip_unindexed<I, A, B, CA, CB>(pi: I, left: CA, right: CB) -> (CA::Result, CB::Result)
+    where I: ParallelIterator<Item = (A, B)>,
+          CA: UnindexedConsumer<A>,
+          CB: UnindexedConsumer<B>,
+          A: Send,
+          B: Send
+{
+    let consumer = UnzipConsumer {
+        left: left,
+        right: right,
+    };
+    pi.drive_unindexed(consumer)
+}
+
+
+/// A fake iterator to intercept the `Consumer` for type `A`.
+struct UnzipA<'b, I, FromB: 'b> {
+    base: I,
+    b: &'b mut FromB,
+}
+
+impl<'b, I, A, B, FromB> ParallelIterator for UnzipA<'b, I, FromB>
+    where I: ParallelIterator<Item = (A, B)>,
+          FromB: Default + ParallelExtend<B>,
+          A: Send,
+          B: Send
+{
+    type Item = A;
+
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+        where C: UnindexedConsumer<Self::Item>
+    {
+        let mut result = None;
+        {
+            // Now it's time to find the consumer for type `B`
+            let iter = UnzipB {
+                base: self.base,
+                left_consumer: consumer,
+                left_result: &mut result,
+            };
+            self.b.par_extend(iter);
+        }
+        result.unwrap()
+    }
+
+    fn opt_len(&mut self) -> Option<usize> {
+        self.base.opt_len()
+    }
```

Comment:
This one got me thinking.  It's fine to leak the source-iterator's index-ness through `opt_len()` here, and it should result in the optimization we want.  However, with an eye on moving that mechanism to specialization, I can't actually implement a true `IndexedParallelIterator` for this!

I could implement `IndexedParallelIterator::drive()`, but not `with_producer()` because of the iterator requirement in `Producer`.  We need to stay in "push" mode here to feed values into multiple consumers, but once we flip to an iterator we're left pulling.  And that means we can't support `zip` et al. on this -- which I wouldn't expect in a `par_extend()` implementation anyway, but with specialization one *could* encounter that.

Maybe this is a distinction which deserves to split up the traits again.  An `IndexedParallelIterator` that can `drive()`, `enumerate()`, etc., and a subset `ProducableParallelIterator` that supports `with_producer()`, `zip()`, etc.

Ugh.

(Incidentally, this "push vs. pull" is basically why `Iterator::unzip` requires `Extend` instead of `FromIterator`.)